### PR TITLE
[FLINK-4297][yarn] decode URL encoded fat jar path

### PR DESCRIPTION
**Pull request by mxm**
This solves problems with spaces and special characters in the
automatically determined fat jar path which is returned URL encoded.

### CODE REVIEW

**Comment by uce** [comment]
Tested this with Hadoop 2.4.1 and it works as expected with `bin/flink run -m yarn-cluster`:
```
2016-08-01 16:32:06,214 INFO  org.apache.flink.yarn.Utils - Copying from 2016-08-01 16:32:06,214 INFO  org.apache.flink.yarn.Utils - Copying from file:/home/hadoop/mxm/flink/flink-dist/target/flink-1.1-SNAPSHOT-bin/flink with spaces/conf/log4j.properties to gs://.../user/hadoop/.flink/application_1470041666362_0007/log4j.properties
...
08/01/2016 16:33:24	Job execution switched to status FINISHED.
```

But YARN sessions don't work. I get:
```
$ bin/yarn-session.sh
Error: Could not find or load main class with
```

**Comment by mxm** [comment]
Thanks for trying it out @uce! There was an unrelated quoting issue in `yarn-session.sh`. I pushed another fix.

**Comment by uce** [comment]
It didn't fix it unfortunately.

**Comment by mxm** [comment]
Should be good to go now.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 1]
```diff
@@ -260,8 +263,15 @@ public AbstractYarnClusterDescriptor createDescriptor(String defaultApplicationN
 		} else {
 			LOG.info("No path for the flink jar passed. Using the location of "
 				+ yarnClusterDescriptor.getClass() + " to locate the jar");
-			localJarPath = new Path("file://" +
-				yarnClusterDescriptor.getClass().getProtectionDomain().getCodeSource().getLocation().getPath());
+			String encodedJarPath =
+				yarnClusterDescriptor.getClass().getProtectionDomain().getCodeSource().getLocation().getPath();
+			try {
+				// we have to decode the url encoded parts of the path
+				localJarPath = new Path("file://" + URLDecoder.decode(encodedJarPath, Charset.defaultCharset().name()));
```
Comment (StephanEwen) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: I think manually coding the `file:// + path` URIs have cross-platform issues.

For File URIs, I found it to be best to go with 
```java
new File(string).toURI()
```
or in that specific case:
```java
localJarPath = new Path(new File(URLDecoder.decode(encodedJarPath, Charset.defaultCharset().name())).toURI());
or
localJarPath = new Path(new URI("file", null, URLDecoder.decode(encodedJarPath, Charset.defaultCharset().name()), null));
```

**Comment by StephanEwen** [comment]
One comment inline, otherwise looks good.

**Comment by StephanEwen** [comment]
Will address the comment and merge this...

**Pull request closed without merging**